File: Lab2/Lab2Sharp/Lab2Sharp/FileManager.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace Lab2Sharp
{
    internal class FileManager
    {
        public void CreateFile(string pathToFile) //створення файлу
        {
            FileStream fileStream = new FileStream(pathToFile, FileMode.Create);
            fileStream.Close();
        }

        public void CreateFile(string pathToFile, string text) //створення файлу
        {
            FileStream fileStream = new FileStream(pathToFile, FileMode.Create);
            byte[] array = Encoding.Default.GetBytes(text);
            fileStream.Write(array, 0, array.Length);
            fileStream.Close();
        }

        public void AppendToFile(string pathToFile, string text) //доповнення файлу
        {
            FileStream fileStream = new FileStream(pathToFile, FileMode.Append);
            byte[] array = Encoding.Default.GetBytes(text + "\n");
            fileStream.Write(array, 0, array.Length);
            fileStream.Close();
        }

        public static string ReadAllFromFile(string pathToFile) //зчитування з файлу
        {
            FileStream fileStream = new FileStream(pathToFile, FileMode.Open);
            byte[] array = new byte[fileStream.Length];
            fileStream.Read(array, 0, array.Length);
            string textFromFile = Encoding.Default.GetString(array);
            fileStream.Close();
            return textFromFile;
        }

        public void PrintAllFile(string pathToFile) //вивід з файлу
        {
            Console.WriteLine($"\nFile name is: {pathToFile.Substring(pathToFile.LastIndexOf('\\') + 1)}");
            string[] text = ReadAllFromFile(pathToFile).Split('\n', StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < text.Length; i++)
            {
                if ((i + 1) % 3 == 0)
                {
                    Console.WriteLine("Gadget");
                    Console.WriteLine($"Name: {text[i - 2]}");
                    Console.WriteLine($"Purchase date: {text[i - 1]}");
                    Console.WriteLine($"Warranty time: {text[i]}");
                    Console.WriteLine();
                }
            }
        }

        public void FillFileFromConsole(string pathToFile) //заповнення файлу з консолі
        {
            Console.Write("Enter number of gadgets: ");
            int gadgetCount = Convert.ToInt32(Console.ReadLine());
            var gadgets = new List<Gadget>();
            Gadget gadget = new Gadget();

            for (int i = 0; i < gadgetCount; i++)
            {
                Console.Write("Enter the name of gadget: ");
                string? inputName = Console.ReadLine();
                Console.Write("Enter gadget's date of purchase (dd.mm.yyyy): ");
                bool correctDate = DateTime.TryParse(Console.ReadLine(), out gadget.PurchaseDate);
                Console.Write("Enter gadget's warranty time (days): ");
                bool correctWarranty = uint.TryParse(Console.ReadLine(), out gadget.WarrantyTime);

                if (inputName != null && correctDate && correctWarranty)
                {
                    gadget.Name = inputName;
                    gadgets.Add(gadget);
                }

                Console.WriteLine();
            }

            CreateFile(pathToFile, GadgetManager.ConvertGadgetsToText(gadgets.ToArray()));
        }
    }
}

File: Lab2/Lab2Python/file_manager.py
import datetime
from datetime import timedelta


def create_file(path, text: str):  # створення файлу
    with open(path, "wb") as file:
        byte_text = str.encode(text)
        file.write(byte_text)


def append_to_file(path, text: str):  # доповнення файлу
    with open(path, "ab") as file:
        byte_text = str.encode(text)
        file.write(byte_text)


def read_all_from_file(path) -> str:  # зчитування з файлу
    byte_array = []
    with open(path, "rb") as file:
        byte_array = file.read()
    return byte_array.decode()


def print_all_file(path: str):  # вивід з файлу
    file_name = path[path.rindex("\\") + 1:]
    print(f"\nFile name is {file_name}")
    text = read_all_from_file(path).split('\n')
    for i in range(len(text)):
        if (i + 1) % 3 == 0:
            print("Gadget")
            print(f"Name: {text[i - 2]}")
            print(f"Purchase date: {text[i - 1]}")
            print(f"Warranty time: {text[i]}")
            print()


def fill_file_from_console(path):  # заповнення файлу з консолі
    gadget_count = int(input("Enter number of gadgets: "))
    gadgets = []
    for i in range(gadget_count):
        gadget_name = input("Enter the name of gadget: ")
        input_date = input("Enter gadget's date of purchase (__.__.____): ")
        purchase_date = datetime.datetime.strptime(input_date, "%d.%m.%Y").date()
        warranty_time = int(input("Enter gadget's warranty time (days): "))
        gadgets.append((gadget_name, purchase_date, warranty_time))
        print()

    create_file(path, convert_gadgets_to_text(gadgets))


def get_gadgets_from_file(path) -> list[tuple]:  # зчитування інформації про гаджети з файлу
    text_from_file = read_all_from_file(path).split('\n')
    gadgets = []

    for i in range(len(text_from_file)):
        if (i + 1) % 3 == 0:
            gadget_name = text_from_file[i - 2]
            purchase_date = datetime.datetime.strptime(text_from_file[i - 1], "%Y-%m-%d").date()
            warranty_time = int(text_from_file[i])
            gadgets.append((gadget_name, purchase_date, warranty_time))

    return gadgets


def get_not_warranty_gadgets(gadgets: list[tuple]) -> list[tuple]:  # отримати гаджети не на гарантії
    return [gadget for gadget in gadgets if gadget[1] + timedelta(gadget[2]) < datetime.datetime.today().date()]


def convert_gadgets_to_text(gadgets: list[tuple]) -> str:  # перетворити інформацію про гаджети в текст
    text = ''
    for gadget in gadgets:
        text += f"{gadget[0]}\n{gadget[1]}\n{gadget[2]}\n"
    return text
